plugins/single_plugins/idle.cpp

```diff
@@ -4,69 +4,297 @@ extern "C"
             return;
 
-        timeout = wlr_idle_timeout_create(wf::get_core().protocols.idle,
+        timeout_dpms = wlr_idle_timeout_create(wf::get_core().protocols.idle,
             wf::get_core().get_current_seat(), 1000 * timeout_sec);
 
-        on_idle.set_callback([&] (void*) {
+        on_idle_dpms.set_callback([&] (void*) {
             set_state(wf::OUTPUT_IMAGE_SOURCE_SELF, wf::OUTPUT_IMAGE_SOURCE_DPMS);
         });
-        on_idle.connect(&timeout->events.idle);
+        on_idle_dpms.connect(&timeout_dpms->events.idle);
 
-        on_resume.set_callback([&] (void*) {
+        on_resume_dpms.set_callback([&] (void*) {
             set_state(wf::OUTPUT_IMAGE_SOURCE_DPMS, wf::OUTPUT_IMAGE_SOURCE_SELF);
         });
-        on_resume.connect(&timeout->events.resume);
+        on_resume_dpms.connect(&timeout_dpms->events.resume);
+    }
+
+    void create_screensaver_timeout(int timeout_sec)
+    {
+        destroy_screensaver_timeout();
+        if (timeout_sec <= 0)
+            return;
+
+        timeout_screensaver = wlr_idle_timeout_create(wf::get_core().protocols.idle,
+            wf::get_core().get_current_seat(), 1000 * timeout_sec);
+        on_idle_screensaver.set_callback([&] (void*) {
+            start_screensaver();
+        });
+        on_idle_screensaver.connect(&timeout_screensaver->events.idle);
+
+        on_resume_screensaver.set_callback([&] (void*) {
+            stop_screensaver();
+        });
+        on_resume_screensaver.connect(&timeout_screensaver->events.resume);
+    }
+
+    void inhibit_outputs()
+    {
+        if (state == SCREENSAVER_DISABLED)
+            return;
+
+        for (auto& output : wf::get_core().output_layout->get_outputs())
+        {
+            if (hook_set[output])
+            {
+                output->render->rem_effect(&render_hook);
+                hook_set[output] = false;
+            }
+            output->render->add_inhibit(true);
+            output->render->damage_whole();
+        }
+        state = SCREENSAVER_DISABLED;
+    }
+
+    void screensaver_frame()
+    {
+        cube_control_signal data;
+        bool all_outputs_active = true;
+
+        if (state == SCREENSAVER_STOPPING && !duration.running())
+        {
+            cube_control_signal data;
+            data.angle = 0.0;
+            data.zoom = ZOOM_BASE;
+            data.last_frame = true;
+            for (auto& output : wf::get_core().output_layout->get_outputs())
+            {
+                output->emit_signal("cube-control", &data);
+                if (hook_set[output])
+                {
+                    output->render->rem_effect(&render_hook);
+                    hook_set[output] = false;
+                }
+                if (state == SCREENSAVER_DISABLED)
+                {
+                    output->render->add_inhibit(false);
+                    output->render->damage_whole();
+                }
+            }
+            state = SCREENSAVER_DISABLED;
```

Comment:
Guess we should clear the map at this point.

Reply:
What do you mean by clear the map? If you mean set all `hook_set` elements to false, this already done in the loop above this line.

Reply:
Oh hm I meant map.clear() so that we don't keep pointers which may become dangling...

Reply:
Ok.